From 0a27c707eafe57a85266452a395f89c8237efdbf Mon Sep 17 00:00:00 2001
From: Jason Gerecke <killertofu@gmail.com>
Date: Mon, 24 Sep 2012 17:53:49 -0700
Subject: [PATCH] Fix double-free in wcmPreInitParseOptions

The 'tool' function here is set equal to 'priv->tool' earlier in
this function. If an error condition occurs, the tool can be freed
both here and in wcmFree (via wcmUninit).

It should be safe to postpone this free() until wcmFree to be
freed along with the rest of the memory allocated in wcmAllocate.
There should be no leak (wcmUninit will be called on failure) and
any code relying on priv->tool being a valid pointer after the
failure of wcmPreInitParseOptions is already broken.

https://bugs.freedesktop.org/show_bug.cgi?id=55200

Signed-off-by: Jason Gerecke <killertofu@gmail.com>
Reviewed-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 src/wcmValidateDevice.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/wcmValidateDevice.c b/src/wcmValidateDevice.c
index 748007b..a296984 100644
--- a/src/wcmValidateDevice.c
+++ b/src/wcmValidateDevice.c
@@ -934,7 +934,6 @@ Bool wcmPreInitParseOptions(InputInfoPtr pInfo, Bool is_primary,
 
 	return TRUE;
 error:
-	free(tool);
 	return FALSE;
 }
 
-- 
1.7.11.4

